class.jetpack.php

```diff
@@ -4019,6 +4022,45 @@ function plugin_action_links( $actions ) {
 		return array_merge( $jetpack_home, $actions );
 	}
 
+	/**
+	 * Filters the login URL to include the registration flow in case the user isn't logged in.
+	 *
+	 * @param string $login_url The wp-login URL.
+	 * @param string $redirect  URL to redirect users after logging in.
+	 * @since Jetpack 8.4
+	 * @return string
+	 */
+	public function login_url( $login_url, $redirect ) {
+		parse_str( wp_parse_url( $redirect, PHP_URL_QUERY ), $redirect_parts );
+		if ( ! empty( $redirect_parts['connect_login_redirect'] ) ) {
+			$login_url = add_query_arg( 'connect_login_redirect', 'true', $login_url );
+		}
+		return $login_url;
+	}
+
+	/**
+	 * Redirects non-authenticated users to authenticate with Calypso if redirect flag is set.
+	 *
+	 * @since Jetpack 8.4
+	 */
+	public function login_init() {
+		// phpcs:ignore WordPress.Security.NonceVerification
+		if ( ! empty( $_GET['connect_login_redirect'] ) ) {
+			add_filter( 'allowed_redirect_hosts', array( &$this, 'allow_wpcom_environments' ) );
+			wp_safe_redirect(
+				add_query_arg(
+					array(
+						'forceInstall' => 1,
+						'url'          => rawurlencode( get_site_url() ),
+					),
+					// @todo provide way to go to specific calypso env.
+					'https://wordpress.com/jetpack/connect'
```

Comment:
There is a `Jetpack::get_calypso_env()` method that tells you what if what kind of environment we are sending the user to. We use the `CALYPSO_ENV` to tell calypso about the local environment. 

Maybe we can add some sort of helper function that helps us update the URL?

Reply:
Yes, I wasn't able to find one either. I can add something, though I think long-term it may be good to have an endpoint that redirects people based on environment.

Reply:
Resolved in 90c8b1ca91b94fb5e22872b0bbc8ae2df260a72e.